lib/babel/transformation/transformers/es6/parameters.rest.js

```diff
@@ -5,6 +5,64 @@ var t    = require("../../../types");
 
 exports.check = t.isRestElement;
 
+var memberExpressionVisitor = {
+  enter: function (node, parent, scope, state) {
+    var localDeclar = scope.getBindingIdentifier(state.name);
+    if (localDeclar !== state.outerDeclar) return;
+
+    if (t.isFunctionDeclaration(node) || t.isFunctionExpression(node)) {
+      state.isOptimizable = false;
+      this.skip();
+      return;
+    }
+
+    if (!t.isReferencedIdentifier(node, parent, { name: state.name })) return;
+
+    if (parent.type === 'MemberExpression') {
+      var prop = parent.property;
+      if (typeof prop.value === 'number' ||
+          prop.type === 'UnaryExpression' ||
+          prop.type === 'BinaryExpression') {
+        state.candidates.push({ node: node, parent: parent });
+        return;
+      }
+    }
+
+    state.isOptimizable = false;
+    this.stop();
+  }
+};
+
+function optimizeMemberExpression(node, parent, offset) {
+  var newExpr;
+
+  var prop = parent.property;
+  switch (prop.type) {
+  case 'Literal':
+    node.name = 'arguments';
+    prop.value += offset;
+    prop.raw = String(prop.value);
+    break;
+  case 'UnaryExpression': {
+    node.name = 'arguments';
+    newExpr = t.binaryExpression('+', prop, t.literal(offset));
+    parent.property = newExpr;
+    break;
+  }
+  case 'BinaryExpression': {
+    node.name = 'arguments';
+    newExpr = t.binaryExpression('+', prop, t.literal(offset));
+    parent.property = newExpr;
+    break;
+  }
+  default:
+    throw new Error('Unsupported property type');
+  }
+}
+
+function optimizeCandidates(candidates) {
```

Comment:
BinaryExpression is used in

    function m(foo, ...bar) {
      var x = bar[1+1];
      var y = bar[x+5];
    }